### PR TITLE
Insert wrapAll pre-interacton for ethflow orders

### PR DESCRIPTION
**Pull request by josojo**
This PR builds the pre-interaction needed to wrap eth for the ethflow orders.

### Test Plan
This can only be properly tested with e2e tests. Hence, I hope it's okay if I don't add a test. Curious about your opinion.

I tested it like this, but its probably not needed for the review:
```
-- merge the PR that activates indexing:
git merge indexing16

— database setup
docker run -d -e POSTGRES_HOST_AUTH_METHOD=trust -e POSTGRES_USER=`whoami` -p 5432:5432 docker.io/postgres
docker build --tag services-migration -f docker/Dockerfile.migration .
docker run -ti -e FLYWAY_URL="jdbc:postgresql://host.docker.internal/?user="$USER"&password=" -v $PWD/database/sql:/flyway/sql services-migration migrate

— running the order book:
export INFURA_KEY=<Your KEY>  && 
cargo run --bin orderbook -- --node-url "https://goerli.infura.io/v3/$INFURA_KEY" --enable-eip1271-orders

— running autopilot
export INFURA_KEY=<Your KEY>   &&
cargo run --bin autopilot -- --node-url "https://goerli.infura.io/v3/$INFURA_KEY" --skip-event-sync

— running solver
export INFURA_KEY=<Your KEY>  &&
Export SOLVER_PK=<Your PK> &&
cargo run -p solver --  --orderbook-url http://localhost:8080/ \
   --node-url "https://goerli.infura.io/v3/$INFURA_KEY" \
  --solver-account $SOLVER_PK\
  --solvers Naive,Baseline \
  --transaction-strategy PublicMempool \
--transaction-submission-nodes "https://goerli.infura.io/v3/$INFURA_KEY"
```

and then I was placing orders with David's interface from here: 
https://github.com/cowprotocol/cowswap/tree/eth-flow/contract-hooks
and set `REACT_APP_API_URL_STAGING_GOERLI=http://localhost:8080/api` in .env file

### CODE REVIEW

**Review by vkgnosis — APPROVED** [review]
I think we should have an ethflow e2e in this codebase. Not sure if needed right now or later when more functionality has been added.

**Comment by fleupold** [comment]
Sorry to comment here. I was surprised to see that we only change `settlement_encoder.rs` to add a very order specific feature.

I wasn't involved in the larger design, but isn't the set of pre-interactions that are needed something the trade/order should know about itself (rather than putting that logic behind an if/else branch in the combined settlement encoding logic which is already quite huge and hard to reason about)? How will this scale when we add EIP2612 and other custom pre/post-interaction support?

**Comment by josojo** [comment]
Hey Felix,

> but isn't the set of pre-interactions that are needed something the trade/order should know about itself

The background of this pre-interaction is a gas estimation. We bundle all the wrapping of eth for all ethflow-orders. Hence, it's not per trade/order, but rather per settlement. Hence, I do the changes in the settlementencoder.

We were aware that this causes some "code smell",
We discussed it here: 
https://cowservices.slack.com/archives/C0375NV72SC/p1663760166726719
and 
https://github.com/cowprotocol/ethflowcontract/pull/15
but our judgment is that the code smell is worth the gas cost savings.

>How will this scale when we add EIP2612 and other custom pre/post-interaction support?
It is so specific to ethflow orders, it's not related to other pre/post interactions. For EIP2612, we would create different pre-interaction based on more involved logic and then merge the pre-interactions with ethflow pre-interactions.

File: crates/solver/src/settlement/settlement_encoder.rs [hunk 1]
```diff
@@ -404,6 +407,44 @@ impl SettlementEncoder {
...
+                .iter()
+                .filter_map(|order_trade| {
+                    if order_trade.trade.order.metadata.is_ethflow_order {
+                        Some(order_trade.trade.order.metadata.owner)
+                    } else {
+                        None
+                    }
+                })
+                .collect(),
+        );
```
Comment (nlordell) [inline crates/solver/src/settlement/settlement_encoder.rs:437]: nit: I think we shouldn't collect + iterate + get, but instead use `find_map` here:
```rust
        let ethflow_owner: Option<H160> = self.liquidity_order_trades
            .iter()
            .map(|order_trade| &order_trade.trade)
            .chain(
                self.order_trades
                    .iter()
                    .map(|order_trade| &order_trade.trade),
            )
            .find_map(|trade| {
                if trade.order.metadata.is_ethflow_order {
                    Some(trade.order.metadata.owner)
                } else {
                    None
                }
            });
```

**Review by nlordell — COMMENTED** [review]
Small style nit: i.e. I don't think we need to collect all trades into a vector in order to find the first ethflow order, but instead just chain the iterators and find them instead.

**Comment by nlordell** [comment]
> I wasn't involved in the larger design, but isn't the set of pre-interactions that are needed something the trade/order should know about itself (rather than putting that logic behind an if/else branch in the combined settlement encoding logic which is already quite huge and hard to reason about)? How will this scale when we add EIP2612 and other custom pre/post-interaction support?

Agreed. As @josojo mentioned, we talked about this code-smell in Slack and decided to proceed with introducing it. Ideally, EthFlow orders would be created with pre-interactions attached to them instead of requiring special handling for the orders in the `solver`. However, this would be a significant effort, and considering that we don't have a concrete timeline for introducing pre-interactions for orders, would delay shipping of EthFlow orders by a while.

**Comment by fleupold** [comment]
> Hence, it's not per trade/order, but rather per settlement. Hence, I do the changes in the settlement encoder.

Maybe it would be reasonable to introduce the concept of "equality" for interactions (e.g. two unlimited approvals to the same contract could also be combined into a single one). This way an individual order could still return a vector of pre interactions and the settlement_encoder could just combines all of them together (while removing duplicates). 

I'm just brainstorming here, if this is a calculated debt we are willing to take, that's ok.

**Comment by josojo** [comment]
IMO the real technical depth was not the "inserting of the pre-interaction", but the is_ethflow_order flag in the orders. 
I think this PR is not really introducing technical depth as the creation of this interaction can super easily be changed later again. 

 I will prepare the e2e test today, once all PRs for this are merged (.e.g we need this one for the e2e test as well:
 https://github.com/cowprotocol/services/pull/588)

**Pull request closed without merging**